cmake: modules: west: Add function to verify blobs

pdgendt opened the pull request:
Add a cmake helper function to verify if blobs have a valid checksum.

pdgendt committed:
cmake: modules: extensions: Add function to verify binary blobs

Add a cmake helper function to verify if blobs have a valid checksum.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

nordicjm reviewed (commented):
Not having used this feature, I'm not sure I understand the purpose of this, they are hosted on git repos right, so git deals with ensuring the files have been correctly checked out, why is this needed?

tejlmand reviewed (changes requested):
Not blocking because of feature itself, but because `west` is an optional tool.

The ability to use `west` to fetch binary blobs is a great help / convenience to users, but users are also allowed to download such blobs manually (or using a custom tool).

Therefore we should not rely on `west` for verifying a blob's SHA, also current implementation proposal does not present how callers of this function will invoke it.

From: https://docs.zephyrproject.org/latest/develop/modules.html#binary-blobs
> Binary blobs are fetched using [west blobs](https://docs.zephyrproject.org/latest/develop/west/zephyr-cmds.html#west-blobs). If west is [not used](https://docs.zephyrproject.org/latest/develop/modules.html#modules-without-west), they must be downloaded and verified manually.



Another minor observation, it seems to me that PR will make it harder to locally develop a bin-blob and test it with Zephyr before releasing it the public, as the `west_blobs_verify()` will fail the build.

pdgendt commented:
@tejlmand thanks for the review, see my comments below:

> Not blocking because of feature itself, but because `west` is an optional tool.
> 
> The ability to use `west` to fetch binary blobs is a great help / convenience to users, but users are also allowed to download such blobs manually (or using a custom tool).

I agree, but users aren't forced to call this function, it CAN be used if `west` is known to be the build tool.

> Therefore we should not rely on `west` for verifying a blob's SHA, also current implementation proposal does not present how callers of this function will invoke it.

I've added examples in the function description? It can be used in `CMakelists.txt`:

```cmake
# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(hello_world)

include(west)

west_blobs_verify()

target_sources(app PRIVATE src/main.c)
```

> Another minor observation, it seems to me that PR will make it harder to locally develop a bin-blob and test it with Zephyr before releasing it the public, as the `west_blobs_verify()` will fail the build.

I don't see why this should be the case? A local file can either be omitted from the `module.yml` file entirely or the sha256 checksum can be calculated locally and updated in `module.yml`.

pdgendt commented:
@tejlmand I've added a check if `WEST` wasn't set to `WEST-NOTFOUND` and print a warning.

tejlmand commented:
> I agree, but users aren't forced to call this function, it CAN be used if west is known to be the build tool.

which was exactly the reason for my question.
Because if called from a (downstream) sample's CMakeLists.txt file. then correct other assumptions can be made by sample author.

But someone could also call those functions elsewhere, as example from a Zephyr module and thus suddenly enforce `west` as a tool, which is mostly my concern.

> A local file can either be omitted from the module.yml file entirely or the sha256 checksum can be calculated locally and updated in module.yml.

Of course users can, just like they can comment out logic inside `west_blobs_verify()` functions.
But should really be possible to just touch code in question, just like when modifying a source file.

pdgendt commented:
> > I agree, but users aren't forced to call this function, it CAN be used if west is known to be the build tool.
> 
> which was exactly the reason for my question. Because if called from a (downstream) sample's CMakeLists.txt file. then correct other assumptions can be made by sample author.
> 
> But someone could also call those functions elsewhere, as example from a Zephyr module and thus suddenly enforce `west` as a tool, which is mostly my concern.

I see, is the latest addition, where `WEST-NOTFOUND` is checked, sufficient to cover this case?

> > A local file can either be omitted from the module.yml file entirely or the sha256 checksum can be calculated locally and updated in module.yml.
> 
> Of course users can, just like they can comment out logic inside `west_blobs_verify()` functions. But should really be possible to just touch code in question, just like when modifying a source file.

I don't really follow here, what would you like to see changed?

tejlmand commented:
> I don't really follow here, what would you like to see changed?

it was more a comment that users testing locally may just omit the file from module.yml or update its SHA.
I just tried to exemplify why such comment was really not useful to users working with the bin-blobs, as users can do anything (such as modify code) to get around build failures, but they shouldn't really have to resort to such kind of things.

Users should be able to replace a local blob with a version of their choice (for example for debug purposes) without doing extra modification to the workspace, just like users can make local modification to source files.

tejlmand commented:
> I see, is the latest addition, where `WEST-NOTFOUND` is checked, sufficient to cover this case?

Now that the general principle of west being optional, let me take a look at the code itself.

tejlmand commented on cmake/modules/west.cmake:
```diff
@@ -95,3 +95,55 @@ else()
+  cmake_parse_arguments(WEST_BLOBS_VERIFY "" "MODULE" "PATHS" ${ARGN})
+
+  execute_process(
+    COMMAND ${WEST} blobs list ${WEST_BLOBS_VERIFY_MODULE} -f "{status} {path}"
```
```suggestion
    COMMAND ${WEST} blobs list ${WEST_BLOBS_VERIFY_MODULE} --format "{status} {path}"
```

tejlmand commented on cmake/modules/west.cmake:
```diff
@@ -95,3 +95,55 @@ else()
+
+  if(WEST_BLOBS_VERIFY_PATHS)
+    foreach(path ${WEST_BLOBS_VERIFY_PATHS})
+      message(STATUS "Verifying blob \"${path}\"")
```
there are currently 91 bin blobs defined in upstream Zephyr modules.
Printing status for each just becomes noise, compared to more important information.

```suggestion
```

If you prefer to keep the message, then please downgrade it to `VERBOSE` instead of `STATUS`.

tejlmand commented on cmake/modules/west.cmake:
```diff
@@ -95,3 +95,55 @@ else()
+      message(STATUS "Verifying blob \"${path}\"")
+
+      if(NOT "${status}" STREQUAL "A")
+        message(FATAL_ERROR "Blob for path \"${path}\" isn't valid. Update with: west blobs fetch")
```
while this can save certain users from wasted debug time due to out-dated bin blobs, then it should still be possible for users to easily replace a bin blob with a custom one where the sha does not match, for example for development or debug purposes.

So either downgrade to a `WARNING` or provide a mechanism / flag to easily silence the error.

tejlmand commented on cmake/modules/west.cmake:
```diff
@@ -95,3 +95,55 @@ else()
+      list(GET blob 1 path)
+      message(STATUS "Verifying blob \"${path}\"")
+
+      if(NOT "${status}" STREQUAL "A")
```
Perhaps compare against `"M"` so that if a bin blob is missing it doesn't generate an error.
(at least for `binary_blobs_verify()`)

pdgendt replied:
A missing blob is treated the same way as an invalid, which makes more sense IMO.

tejlmand replied:
no, not in all cases.

we should not allow someone to introduce a `binary_blobs_verify()` in module `foo`, thus breaking the build if module `bar` is missing a blob.

A build should only breakif the specific blob required for this build is missing, hence the proposal for the `REQUIRED` together with `PATHS`.
It should not break because some blob, not used in this build, is missing.

Alternative remove the pure `binary_blobs_verify()` and always request `MODULE` or `PATHS`.

As example, if i'm build for infineon platform, the build should not fail because an espressif blob is missing.

Similar, if i'm build for Gecko SoC series `efr32mg12p`, then the build shouldn't fail because `librail_efr32xg13_gcc_release.a` is missing.

A build should only fail if the needed dependencies for that particular build is missing.

pdgendt replied:
I like the idea of a module to only allow verifying its own blobs, but for this to work, I need to get the current module name in `binary_blobs_verify`. Is that possible?

tejlmand replied:
we do have `module_name`, but currently considered build system internals.
If we have a use-case (like we do now), then we should create a corresponding `ZEPHYR_CURRENT_MODULE_NAME` here:
https://github.com/zephyrproject-rtos/zephyr/blob/982fc416f8df34977a720a372564b3b673e1fd1e/CMakeLists.txt#L575-L576

but before doing so, we must ensure that `west blobs` is actually using the Zephyr module's name, and not the `west project` name, as those two names may not be identical.

tejlmand reviewed (changes requested):
Thanks for this proposal.

I do see some positive in the proposal, but we should also be a little careful.
Just like users may choose to mix / use a SHA on a project which is different from the SHA in the west manifest file (without warning / errors), then users are also allowed to use a custom bin blob (older / newer).

Users are responsible for running `west blobs`, just like they must remember to run `west update`.

That said, I know how annoying it is to things not working due to wrong code, such as a wrong bin blob.
So i'm positive to the general idea, but has some comments.

Upstream Zephyr modules currently have 91 bin blobs defined.
The whole idea behind the blobs mechanism was that users should only fetch blobs from the vendors they need, thus having `binary_blobs_verify()` failing a build because one blob is missing is a no-go.
What can be done instead is to only check SHAs of the bin-blobs actually present, that is testing against `M`, instead of `A`.

Another use-case I see is the ability for a module to verify that a specific blob is present before continuing to the build stage, and for this we can adjust the call a bit, like this:
```
binary_blobs_verify(PATHS <file> [REQUIRED])
``` 
where the REQUIRED indicates that verify should fail if file is missing (Current behavior).

But if `REQUIRED` is omitted, then only bin blobs present are verified.

This will allow a module's CMake code to do:
```
binary_blobs_verify(PATHS img/bluetooth/firmware/cy_ble_stack_controller.a REQUIRED)
```
before:
https://github.com/zephyrproject-rtos/zephyr/blob/4204ca9bcb3f675a0e63f647931ffe2c4b515729/modules/hal_infineon/bless/CMakeLists.txt#L20-L23

pdgendt commented:
Thanks @tejlmand for the suggestions, I've updated the PR.

The current implementation allows you to specify paths for binary blobs, but if omitted simply verifies all the blobs. The `MODULE` argument can limit that to a single module.

pdgendt commented:
@tejlmand I've made the `MODULE` argument a required one, I think that is closer to what is expected?

pdgendt committed:
cmake: Introduce ZEPHYR_CURRENT_MODULE_NAME

Add a cmake variable for the current module's name.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
Marking as `DNM` until I have been able to test the module name.

pdgendt committed:
doc: develop: modules: Add entry for ZEPHYR_CURRENT_MODULE_NAME

A new cmake variable is added for zephyr modules, namely
`${ZEPHYR_CURRENT_MODULE_NAME}`. Add it to the documentation.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

tejlmand commented on CMakeLists.txt:
```diff
@@ -581,6 +581,7 @@ foreach(module_name ${ZEPHYR_MODULE_NAMES})
   # https://cmake.org/pipermail/cmake/2019-June/069547.html
   zephyr_string(SANITIZE TOUPPER MODULE_NAME_UPPER ${module_name})
   if(NOT ${ZEPHYR_${MODULE_NAME_UPPER}_CMAKE_DIR} STREQUAL "")
+    set(ZEPHYR_CURRENT_MODULE_NAME ${module_name})
```
Please clear `ZEPHYR_CURRENT_MODULE_NAME` like this:
https://github.com/zephyrproject-rtos/zephyr/blob/7fbcbcbfa7b0563aa16e48d3ae9c762ccd04debc/CMakeLists.txt#L591-L592

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,66 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+# Paths are relative to the current module's zephyr/blobs directory.
+#
+# Usage:
+#   zephyr_blobs_verify([PATHS path [paths...]]
```
please let the caller decide between `MODULES` or `PATHS`.

Also, imo `FILES` are a better name than `PATHS`.

So that we have:
```
zephyr_blobs_verify(<MODULE <name>|FILES file [file...]> [REQUIRED])
```

pdgendt replied:
It is possible to have the same relative path (to module's `zephyr/blobs`') for different modules, which can be a bit tricky if only files are listed?

Should `MODULE` become a required argument, now that `ZEPHYR_CURRENT_MODULE_NAME` is available?

tejlmand replied:
> It is possible to have the same relative path (to module's zephyr/blobs') for different modules, which can be a bit tricky if only files are listed?

which was exactly the reason to support abspath.
A caller outside of a module's cmake code can then use: `${ZEPHYR_<module>_MODULE_DIR}/path/to/file` to set abspath.
A caller inside the module can use a path relative to the current CMakeLists.txt file (which sets CMAKE_CURRENT_SOURCE_DIR when added).

> Should MODULE become a required argument, now that ZEPHYR_CURRENT_MODULE_NAME is available?

No, with abspath support, the module is not needed as required arg.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,66 @@ function(zephyr_syscall_header_ifdef feature_toggle)
 
+# Verify blobs fetched using west. If the sha256 checksum isn't valid, a warning/
+# fatal error message is printed (depends on REQUIRED flag).
+# Paths are relative to the current module's zephyr/blobs directory.
```
Not good.

Please let code accept absolute paths, and when a path is relative, then it should be relative to `CMAKE_CURRENT_SOURCE_DIR` which is most common in CMake.
Will also allow the function to be called outside of a module, for example like:
```
zephyr_blobs_verify(FILES ${ZEPHYR_HAL_INFENION_MODULE_DIR}/zephyr/blobs/img/cat1cm0p/COMPONENT_CM0P_SLEEP/psoc6_01_cm0p_sleep.bin)
```

Or when inside a module's CMakeLists.txt file, developer may do like:
```
zephyr_blobs_verify(FILES
    ${bless_blobs_dir}/COMPONENT_BLESS_CONTROLLER/cy_ble_stack_controller.a
  	${bless_blobs_dir}/COMPONENT_BLESS_CONTROLLER/cy_ble_stack_manager.a
)
```
in a place like here:
https://github.com/zephyrproject-rtos/zephyr/blob/4403d8f4c622fe858a57aeea4b555256102fa890/modules/hal_infineon/bless/CMakeLists.txt#L20-L23

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,66 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+  cmake_parse_arguments(BLOBS_VERIFY "REQUIRED" "" "PATHS" ${ARGN})
+
+  execute_process(
+    COMMAND ${WEST} blobs list ${ZEPHYR_CURRENT_MODULE_NAME} --format "{status} {path}"
```
should probably use `abspath` here, so that the exact location is know.
This will also make abspath and relpath support in CMake code cleaner.

```suggestion
    COMMAND ${WEST} blobs list ${ZEPHYR_CURRENT_MODULE_NAME} --format "{status} {abspath}"
```

tejlmand replied:
also, including `"{module}"` will provide the module where the missing blob originates, making it possible to inform the user the exact module to use when doing `west blobs fetch <module>`.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,66 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+# zephyr_blobs_verify(REQUIRED)           # verify all blobs and fail on error
+# zephyr_blobs_verify(PATHS img/file.bin) # verify a single path
+function(zephyr_blobs_verify)
+  if(NOT DEFINED ZEPHYR_CURRENT_MODULE_NAME)
```
Module should be caller specified, caller may choose to use `MODULE ${CMAKE_CURRENT_MODULE_NAME}` when calling the function from inside its own CMake glue code.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,66 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+
+      # Each path that has a correct sha256 is prefixed with an A
+      if(NOT "A ${path}" IN_LIST BLOBS_LIST)
+        message(${msg_lvl} "Blob for path \"${path}\" isn't valid. Update with: west blobs fetch")
```
Please inform users to just fetch blobs for the module which is failing.

No reason to fetch blobs for a module which may not be used by the user.

tejlmand reviewed (changes requested):
Much better.

Generally still not fond of the `MODULE` being able to break a build when files unrelated to current build is not present or having wrong SHA.

Thus i'm much more in favor of a check like:
```
    zephyr_blob_verify(FILES gecko/platform/radio/rail_lib/autogen/librail_release/librail_efr32xg${GECKO_SERIES_NUMBER}_gcc_release.a)
    add_prebuilt_library(librail      gecko/platform/radio/rail_lib/autogen/librail_release/librail_efr32xg${GECKO_SERIES_NUMBER}_gcc_release.a)
```
Ref: https://github.com/zephyrproject-rtos/hal_silabs/blob/11ab59175a5ded618680a7692dbaae8f4fbd6325/gecko/CMakeLists.txt#L92C1-L92C145

over:
```
    zephyr_blob_verify(MODULE silabs)
    add_prebuilt_library(librail      gecko/platform/radio/rail_lib/autogen/librail_release/librail_efr32xg${GECKO_SERIES_NUMBER}_gcc_release.a)
```

But as the most specific call available for fetching blobs is `west blobs fetch <MODULE>`, then it is accepted to support a `zephyr_blobs_verify(MODULE <module>)` and let the caller decide between the two options.

pdgendt commented:
@tejlmand thanks for the effort, I've updated the PR

As per the suggestions I feel that the `MODULE` argument should be required, it is now possible to pass `${ZEPHYR_CURRENT_MODULE_NAME}` anyway.

pdgendt commented:
I would also like to add a test case, but I'm not sure where to add this.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,71 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+
+  if(DEFINED BLOBS_VERIFY_FILES)
+    foreach(file ${BLOBS_VERIFY_FILES})
+      # Resolve path.
```
We need a call to:
```
file(REAL_PATH ${file} real_path)
file(REAL_PATH ${CMAKE_CURRENT_SOURCE_DIR}/${file} real_path)
```
to ensure any `../` are correctly transformed and can be compared to the output from `blobs list`.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,71 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+    foreach(file ${BLOBS_VERIFY_FILES})
+      # Resolve path.
+      if(IS_ABSOLUTE ${file})
+        file(TO_NATIVE_PATH "${file}" path)
```
Good catch that `blobs list --format "{abspath}"` is actually using platform path separators 👍 .

We can extend `blobs list` format specifier to accept `--format="{posixpath}"` and avoid this call, and another benefit of that is that it will align the function with `west list --format="{path} {abspath} {posixpath}"`.

Note, keeping a `TO_NATIVE_PATH` is also acceptable.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1559,6 +1559,71 @@ function(zephyr_syscall_header_ifdef feature_toggle)
+# fatal error message is printed (depends on REQUIRED flag).
+#
+# Usage:
+#   zephyr_blobs_verify(MODULE module [FILES file [files...]] [REQUIRED])
```
if `FILES` is given, then there is no reason for `MODULE` to be required argument.

A relative path argument is relative to `CMAKE_CURRENT_SOURCE_DIR`, so the module name is not used for path expansion. Also the module name isn't used later in `if(NOT "A ${path}" IN_LIST BLOBS_LIST)` and `{abspath}` ensures that all blobs have unique paths.

fabiobaltieri commented:
Hi, the CI failure has been fixed upstream, if you rebase it should pass.

tejlmand reviewed (approved):
thanks for the effort and patience 👍

pdgendt commented:
@carlescufi you mind taking a look at this PR, as you've introduced the blobs infrastructure? Thanks
